internal/plugin/connectors/tcp/mssql/connector.go

```diff
@@ -159,99 +164,121 @@ func (connector *SingleUseConnector) Connect(
 	// Create the context for our connection
 	ctx := context.Background()
 
-	// Create a channel for receiving the prelogin response through the context
-	preLoginResponseChannel := make(chan map[uint8][]byte)
-
-	// Create a channel for send the client login to the driver through the context
-	clientLoginChannel := make(chan mssql.Login)
-
 	// Set a 'marker' for when the driver has finished connecting to the server
-	connectPhaseFinished := make(chan struct{})
+	connectionResultChan := make(chan connectResult)
 
-	// Add channels to the context, to exchange information with the driver
-	loginContext := context.WithValue(
-		ctx,
-		ctxtypes.PreLoginResponseKey,
-		preLoginResponseChannel)
-	loginContext = context.WithValue(
-		loginContext,
-		ctxtypes.ClientLoginKey,
-		clientLoginChannel)
+	// Create a ConnectIntercept for exchanging values with the driver via context and
+	// ensure it gets cleaned up when this method returns
+	connInterceptor := mssql.NewConnectInterceptor()
 
-	// Build a new driver connection
-	var driverConn driver.Conn
+	// Add connInterceptor to the context for exchanging values with the driver
+	loginContext := mssql.NewContextWithConnectInterceptor(ctx, connInterceptor)
 
 	go func() {
 		// Kick off authentication through our third party connector
-		// TODO: this is generally dangerous
-		//  we're sharing state between the main routine and this, and have the channel
-		//  working as an implicit lock
-		driverConn, err = driverConnector.Connect(loginContext)
-		connectPhaseFinished <- struct{}{}
-	} ()
-
-	// Blocks continuation until we've received the preLoginResponse from the driver
-	preloginResponse := <- preLoginResponseChannel
+		driverConn, err := driverConnector.Connect(loginContext)
 
-	// Since the communication between the client and Secretless must be unencrypted,
-	// we fool the client into thinking that it's talking to a server that does not support
-	// encryption.
-	preloginResponse[mssql.PreloginENCRYPTION] = []byte{mssql.EncryptNotSup}
-
-	// Write the prelogin packet back to the user
-	err = mssql.WritePreloginWithPacketType(clientBuffer, preloginResponse, mssql.PackReply)
-	if err != nil {
-		wrappedError := errors.Wrap(err, "failed to write prelogin response")
-		connector.sendError(wrappedError)
-		return nil, wrappedError
-	}
+		connectionResultChan <-connectResult{
+			conn: driverConn.(*mssql.Conn),
+			err:  err,
+		}
+	} ()
 
-	// We parse the client's Login packet so that we can pass on params to the server.
-	clientLogin, err := mssql.ReadLogin(clientBuffer)
-	if err != nil {
-		wrappedError := errors.Wrap(err, "failed to handle client login")
-		connector.sendError(wrappedError)
-		return nil, wrappedError
+	var loginAck mssql.LoginAckStruct
+	// Block continuation until error or driver has completed connection
+responseLoop:	for {
+		select {
+		// preLoginResponse is received from the server
+		case preloginResponse := <-connInterceptor.PreLoginResponse:
```

Comment:
Looking at this a second (or third?) time... The code before the select does a call to driverConnector.Connect() and then sends a connection result in the connectionResultChan. We're assuming that a preloginResponse will be received before the result is seen in the connectionResultChan, but there could be a race. Should we defer the sending of the connection result, or defer the checking for it, or something like that to ensure we're processing things in the required order?

Reply:
@diverdane It looks to me like line 181 can't be hit until (among other things) the internal code of `driverConnector.Connect` puts a value on the preloginResponse channel _and_ that value is pulled off (which can only happen via line 192).  So there should be no race.

That said, the fact that this is confusing enough that you read it that way does cement my believe we should flatten the logic here (though I'm still fine with doing that in a followon)

Reply:
Ah, right. I think I realized the first time that I reviewed this part that there was synchronization built in.